File: Java-Core/src/main/java/presto/grammar/UnresolvedArgument.java
package presto.grammar;

import presto.declaration.AttributeDeclaration;
import presto.declaration.IDeclaration;
import presto.error.PrestoError;
import presto.error.SyntaxError;
import presto.expression.IExpression;
import presto.parser.Dialect;
import presto.runtime.Context;
import presto.runtime.Context.MethodDeclarationMap;
import presto.type.IType;
import presto.utils.CodeWriter;
import presto.value.IValue;

public class UnresolvedArgument extends BaseArgument implements INamedArgument {

	INamedArgument resolved = null;
	
	public UnresolvedArgument(Identifier name) {
		super(name);
	}
	
	@Override
	public String getSignature(Dialect dialect) {
		return getName().toString();
	}

	@Override
	public void toDialect(CodeWriter writer) {
		writer.append(name);
		if(defaultExpression!=null) {
			writer.append(" = ");
			defaultExpression.toDialect(writer);
		}
	}
	
	@Override
	public String toString() {
		return name.toString();
	}
	
	@Override
	public void check(Context context) throws SyntaxError {
		resolveAndCheck(context);
	}
	
	@Override
	public String getProto(Context context) throws SyntaxError {
		resolveAndCheck(context);
		return resolved.getProto(context);
	}
	
	@Override
	public IType getType(Context context) throws SyntaxError {
		resolveAndCheck(context);
		return resolved.getType(context);
	}
	
	@Override
	public void register(Context context) throws SyntaxError {
		resolveAndCheck(context);
		resolved.register(context);
	}
	
	@Override
	public IValue checkValue(Context context, IExpression value) throws PrestoError {
		resolveAndCheck(context);
		return resolved.checkValue(context, value);
	}
	
	private void resolveAndCheck(Context context) throws SyntaxError {
		if(resolved!=null) 
			return;
		IDeclaration named = context.getRegisteredDeclaration(IDeclaration.class, name);
		if(named instanceof AttributeDeclaration)
			resolved = new AttributeArgument(name);
		else if(named instanceof MethodDeclarationMap)
			resolved = new MethodArgument(name);
		else
			throw new SyntaxError("Unknown identifier:" + name);
	}


}

File: Java-CodeServer/src/main/resources/js/prompto/grammar/UnresolvedArgument.js
var SyntaxError = require("../error/SyntaxError").SyntaxError;
var AttributeDeclaration = require("../declaration/AttributeDeclaration").AttributeDeclaration;
var AttributeArgument = require("./AttributeArgument").AttributeArgument;
var MethodDeclarationMap = require("../runtime/Context").MethodDeclarationMap;
var MethodArgument = require("./MethodArgument").MethodArgument;

function UnresolvedArgument(name) {
	this.name = name;
	this.resolved = null;
	return this;
}

/*
@Override
public String getSignature(Dialect dialect) {
	return getName();
}

@Override
public String getName() {
	return name;
}

@Override
public String toString() {
	return name;
}
*/

UnresolvedArgument.prototype.toDialect = function(writer) {
    writer.append(this.name);
};

UnresolvedArgument.prototype.check = function(context) {
	this.resolveAndCheck(context);
};

UnresolvedArgument.prototype.getProto = function(context) {
	this.resolveAndCheck(context);
	return this.resolved.getProto(context);
};

UnresolvedArgument.prototype.getType = function(context) {
	this.resolveAndCheck(context);
	return this.resolved.getType(context);
};

UnresolvedArgument.prototype.register = function(context) {
	this.resolveAndCheck(context);
	this.resolved.register(context);
    if(this.defaultExpression!=null)
        context.setValue(this.name, this.defaultExpression.interpret(context));
};

UnresolvedArgument.prototype.checkValue = function(context, value) {
	this.resolveAndCheck(context);
	return this.resolved.checkValue(context, value);
}

UnresolvedArgument.prototype.resolveAndCheck = function(context) {
	if(this.resolved!=null) {
		return;
	}
	var named = context.getRegisteredDeclaration(this.name);
	if(named instanceof AttributeDeclaration) {
		this.resolved = new AttributeArgument(this.name);
	} else if(named instanceof MethodDeclarationMap) {
		this.resolved = new MethodArgument(this.name);
	} else {
		throw new SyntaxError("Unknown identifier:" + this.name);
	}
};

exports.UnresolvedArgument = UnresolvedArgument;

